dynet/expr.h

```diff
@@ -35,17 +36,38 @@ struct Expression {
   VariableIndex i;
   unsigned graph_id;
 
-  Expression() : pg(nullptr), i(0), graph_id(0) { }
-  const bool is_stale() const {return (get_number_of_active_graphs() != 1 || graph_id != get_current_graph_id());}
+  Expression() : pg(nullptr), i(0), graph_id(0) {}
+
+  Expression(Device *device) : pg(nullptr), i(0), graph_id(0) {}
   /**
    * \brief Base expression constructor
    * \details Used when creating operations
    *
    * \param pg Pointer to the computation graph
    * \param i Variable index
-   * \param name Name of the expression
    */
-  Expression(ComputationGraph *pg, VariableIndex i) : pg(pg), i(i), graph_id(pg->get_id()) { }
+  Expression(ComputationGraph *pg, VariableIndex i) : pg(pg),
+    i(i), graph_id(pg->get_id()) {}
+  /**
+   * \brief Base expression constructor
+   * \details Used when creating operations
+   *
+   * \param pg Pointer to the computation graph
+   * \param i Variable index
+   * \param device Device placement for the Expression
+   */
+  Expression(ComputationGraph *pg, VariableIndex i, Device *device) : pg(pg),
+    i(i), graph_id(pg->get_id()) {}
+
+  inline std::string get_device() const { return pg->nodes[i]->device->name; }
```

Comment:
I don't understand what you mean here. My comment was just that you should change the name of the function, and also make sure that if we call this when `pg->nodes[i]->device` is null, we should get an error instead of a segmentation fault.